twister: Add quarantine feature

PerMac opened the pull request:
twister: Add quarantine feature

Adds feature allowing to use yaml file with dictionaries defining
tests to be quarantined (extra arg "--quarantine-list FILENAME").
The dictionaries are validated according to the proper schema
and loaded.

A flat list is created containing quarantined configurations
(configuration = platform + scenario). Configurations under quarantine
are skipped and get "Quarantine" as a reason in the results reports.

A "comment" can be added to a quarantine entry in the quarantine yaml
with more details (e.g. issue #) and it will be also added to
the report.

The status of tests under quarantine can be verify if
`--quarantine-verify` is used in addition to
"--quarantine-list FILENAME". Using these args will make twister skip
all tests which are not on the quarantine list.

Fixes #33207

PerMac committed:
twister: Add quarantine feature

Adds feature allowing to use yaml file with dictionaries defining
tests to be quarantined (extra arg "--quarantine-list FILENAME").
The dictionaries are validated according to the proper schema
and loaded.

A flat list is created containing quarantined configurations
(configuration = platform + scenario). Configurations under quarantine
are skipped and get "Quarantine" as a reason in the results reports.

A "comment" can be added to a quarantine entry in the quarantine yaml
with more details (e.g. issue #) and it will be also added to
the report.

The status of tests under quarantine can be verify if
`--quarantine-verify` is used in addition to
"--quarantine-list FILENAME". Using these args will make twister skip
all tests which are not on the quarantine list.

Signed-off-by: Maciej Perkowski <Maciej.Perkowski@nordicsemi.no>

PerMac commented:
To try run:
```
scripts/twister -T samples/hello_world/ -T tests/posix/eventfd_basic/ -T tests/kernel/common/ -p qemu_cortex_m3 -p native_posix --platform-reports -q quarantine.yaml
```
and check the result reports.
@andyross @galak @hakehuang @erwango @chen-png I've marked you here as I thought you might be interested in that feature and want to play a bit with it. I would appreciate your input. I believe it can be really useful for on-target CIs and results publishing. It would allow to easily "turn off" the failing tests and got it marked in the results as "Quarantine" so it won't block the publishing due to hitting the failure threshold. It should also improve the execution time, as faulty tests won't be executed and `retry-failed` won't consume the time. 
@nashif It also should help in results visualization in grafana, where only "Quarantine" skips can be counted for pie charts and pass/fail/skip ratios 
The proposed implementation leaves some open space for expanding the functionality. I didn't add an option to skip the whole test suite (basically all scenarios in yaml) as there is a twister refactoring going on and it is better to wait till then than implementing some temp workarounds.) 
If the proposed solution is accepted I will add the proper documentation and the example of quarantine.yaml will end up there.

nashif commented:
thanks, will take a look. 

unrelated to the content of the PR, are you adding the related issue directly on the right sidebar? Can you please instead use the 'Fixes #XYZ' in the PR body?

hakehuang commented on quarantine.yaml:
```diff
@@ -0,0 +1,20 @@
+list1:
+  platforms: all
+  scenarios:
+    - sample.basic.helloworld
```
suggest to add a github issue link here

hakehuang commented on scripts/pylib/twister/twisterlib.py:
```diff
@@ -3074,6 +3108,11 @@ def apply_filters(self, **kwargs):
+                                               instance.testcase.id])
+                if test_configuration in self.quarantine:
+                    discards[instance] = discards.get(instance,
+                                                      "Quarantine")
```
just wonder whether it is possible to add the github issue link into report, so that we can tarck the latest status

PerMac replied:
It is a good idea to be able to easily track related bugs. The simplest, for now, is to add comments into quarantine.yaml with the related links and whatever information the user wants. Is the link useful in the report? We can discuss this during the testing wg meeting.

hakehuang reviewed (approved):
just have one enhancement comments, looks good to me

chen-png reviewed (approved):
tested and it worked well.

evgeniy-paltsev commented on scripts/twister:
```diff
@@ -464,6 +464,15 @@ Artificially long but functional example:
         help="Load list of tests and platforms to be run from file.")
 
+    parser.add_argument(
+        "-q",
```
I guess it's better not to use '-q' for this purpose as usually '-q' stands for 'quiet'. So it's quite misleading in this case.

PerMac replied:
I removed `-q`, only full `--quarantine-list` will work now

hakehuang commented:
- Need to explain why cases are in quarantine list, maybe add a comment with link to github issue would be a solution.
- Need to document the quarantine yaml schema.
- Maybe we can use this quarantine list file as input for a checking test plan. Such as all the excluded cases can be checked whether runnable or not.
- Quarantine cases may not be reported as skipped, should be a special <quarantine message=”” type=””> currently it is <skipped message=”quarantine” type=”skip”>

evgeniy-paltsev commented on quarantine.yaml:
```diff
@@ -0,0 +1,20 @@
+# list of quarantined tests
```
Just a nitpick - probably we can move `quarantine.yaml` to `tests/` directory so we won't overwhelm zephyr home directory?

PerMac replied:
This file will be removed from this dir and will probably end in docs as an example. For now, it just serves as a showcase to play with for the review purpose.

nashif commented on quarantine.yaml:
```diff
@@ -0,0 +1,20 @@
+# list of quarantined tests
+
+list1:
```
maybe move to a list:

```
- platforms:
    - native_posix
  scenarios:
    - posix.eventfd_basic.posix_api
- platforms:
    - qemu_cortex_m3
    - native_posix
  scenarios:
    - kernel.common
    - kernel.common.misra
    - kernel.common.nano64
```

and add schema verification as well

nashif commented on quarantine.yaml:
```diff
@@ -0,0 +1,20 @@
+  platforms:
+    - native_posix
+  scenarios:
+    - posix.eventfd_basic.posix_api
```
add a details section for documenting why something is being quarantined.

PerMac commented:
@nashif @hakehuang @chen-png  Can you please check again this PR? I've added the requested changes/features:
- comments can be added in the quarantine list (e.g. link to PR) which will also end up in the report
- additional arg `--quarantine-verify` can be used which will make twister filter out everything which is not in the quarantine list (to verify only the statuses of quarantine items)
- I added a schema verification for the .yaml file
I made separate commits for this changes so it is easier to follow what have changed. In the end, I will squash some of them.
The `quarantine.yaml` will be removed and will end up in docs as an example of how to use the quarantine. 
I will add docs when it is agreed that the proposed solution is fine

PerMac commented:
I realized that in one of the commits I accidentally make pycharm refactor automatically all the twisterlib.py. I'm fixing this now

nashif commented:
if this is ready for review, move it out of draft please.

PerMac committed:
docs: twister: Add section about quarantine feature

Add section about quarantine feature

Signed-off-by: Maciej Perkowski <Maciej.Perkowski@nordicsemi.no>

PerMac commented:
@nashif ready for review

nashif commented on doc/guides/test/twister.rst:
```diff
@@ -634,3 +634,39 @@ using an external J-Link probe.  The "probe_id" keyword overrides the
       serial: null
+
+Quarantine
++++++++++++++++++++++++++++
```
make this match length of title and add space below.

PerMac replied:
fixed

nashif reviewed (changes requested):
minor, otherwise lgtm
